projects/graph/src/graph.py

```diff
@@ -2,8 +2,106 @@
 Simple graph implementation compatible with BokehGraph class.
 """
 
+from draw import BokehGraph
 
 class Graph:
     """Represent a graph as a dictionary of vertices mapping labels to edges."""
+
     def __init__(self):
-        pass  # TODO
+        self.vertices = {}
+
+    def add_edge(self, start, end, bidirectional=True):
+        """Add an edge from start to end."""
+        if start not in self.vertices or end not in self.vertices:
+            raise Exception("Error - vertices not in graph!")
+        self.vertices[start].add(end)
+        if bidirectional:
+            self.vertices[end].add(start)
+    """
+    DFS(graph):
+    for v of graph.verts:
+        v.color = white
+        v.parent = null
+
+    for v of graph.verts:
+        if v.color == white:
+            DFS_visit(v)
+    """
+
+    def dft(self, starting_node, visited=None):
```

Comment:
This was the hardest part of the week but, it's also the single most important part of what we've studied so far in the CS portion of the curriculum. I would highly recommend you go back and study this until it's ingrained in your mind just as much as create-react-app is. It will mean the difference between getting a junior position or a senior position later on in your career.